File: 2 Three Number Sum/ThreeNumSum.cpp
#include <iostream>
#include <vector>

using namespace std;

// O(n^2)
vector<vector<int>> threeNumberSum(vector<int> array, int targetSum){
	// O(log(n))
	sort(array.begin(), array.end());
	int curr = 0;
	vector<vector<int>>triplets;
	
	// O(n)
	while(curr < array.size() - 2){
    		int left = curr + 1;
    		int right = array.size() - 1;
		// O(n)
		while(left != right){
			int sum = array[curr] + array[left] + array[right];
			if(sum > targetSum)
				right--;
			else if (sum < targetSum)
				left++;
			else
				triplets.push_back({array[curr], array[left++], array[right]});
		}
      		curr++;
	}
	return triplets;
}

File: 2 Three Number Sum/ThreeNumSum.py
'''
    Time: O(nlogn) + O(n^2) = O(n^2)
    Space: O(n)
'''
def threeNumberSum(array, targetSum):
    # sort the array O(nlogn)
    array.sort()

    # list to store sums
    sums = []

    for index, num in enumerate(array):
        leftPtr = index + 1
        rightPtr = len(array) - 1
        while leftPtr < rightPtr:
            potentialSum = num + array[leftPtr] + array[rightPtr]

            if potentialSum == targetSum:
                sums.append([num, array[leftPtr], array[rightPtr]])
                leftPtr += 1
                rightPtr -= 1
            elif potentialSum > targetSum:
                rightPtr -= 1
            else:
                leftPtr += 1
    return sums